### PR TITLE
endpoint: extract endpoint configuration update from `regeneratePolicy` 

### PR DESCRIPTION
**Pull request by ianvernon**
Decoupling policy updates from configuration updates allows for policy computation to be entirely separate from endpoint configuration. As a result, policy generation for an endpoint is now based only upon the endpoint's identity / set of labels in relation to the policy repository. This will make testing endpoint policy generation much easier in the future.

Signed-off by: Ian Vernon <ian@cilium.io>

---
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/5324)

### CODE REVIEW

**Comment by ianvernon** [comment]
test-me-please

**Comment by ianvernon** [comment]
test-me-please

**Comment by ianvernon** [comment]
test-me-please

**Comment by ianvernon** [comment]
test-me-please

**Comment by ianvernon** [comment]
Tests failed due to unused import that I didn't clean up in unit tests... trying again.

**Comment by ianvernon** [comment]
test-me-please

**Comment by ianvernon** [comment]
test-me-please

**Comment by ianvernon** [comment]
test-me-please

File: pkg/endpoint/policy.go [hunk 1]
```diff
@@ -545,6 +525,14 @@ func (e *Endpoint) regeneratePolicy(owner Owner, opts models.ConfigurationMap) (
 		return e.nextPolicyRevision > e.policyRevision, nil
 	}
 
+	e.prevIdentityCache = labelsMap
+
+	// First step when calculating policy is to check whether ingress or egress
+	// policy applies (i.e., if rules select this endpoint). We can use this
+	// information to short-circuit policy generation if enforcement is
+	// disabled for ingress and / or egress.
+	e.ingressPolicyEnabled, e.egressPolicyEnabled = owner.EnableEndpointPolicyEnforcement(e)
```
Comment (ianvernon) [inline pkg/endpoint/policy.go:535]: I have another PR coming up soon which checks these values before generating endpoint policy. We can short-circuit policy-generation in this case.

**Comment by ianvernon** [comment]
test-me-please

**Comment by ianvernon** [comment]
Re-testing. The build got scheduled on a bad node on Jenkins.

**Comment by ianvernon** [comment]
test-me-please

File: pkg/endpoint/policy.go [hunk 2]
```diff
@@ -568,46 +544,46 @@ func (e *Endpoint) regeneratePolicy(owner Owner, opts models.ConfigurationMap) (
...
 		"policyRevision.next": e.nextPolicyRevision,
+		"forcedRegeneration":  forceRegeneration,
 	}).Debug("Done calculating policy")
 
-	needToRegenerateBPF := optsChanged || policyChanged || e.nextPolicyRevision > e.policyRevision
+	// If the policy changed, or the revision of the policy repository has changed
+	// we return true. It is possible that the endpoint's next policy revision
+	// is the same as the endpoint's current policy revision; this indicates
+	// that no new rules have been added in the policy repository; if policy
+	// hasn't changed, and no new rules were added, then
```
Comment (raybejjani) [inline pkg/endpoint/policy.go]: `then return false`, perhaps?

File: pkg/endpoint/bpf.go [hunk 3]
```diff
@@ -465,7 +465,7 @@ func (e *Endpoint) regenerateBPF(owner Owner, epdir, reason string) (revnum uint
 	// In the first ever regeneration of the endpoint, the conntrack table
 	// is cleaned from the new endpoint IPs as it is guaranteed that any
 	// pre-existing connections using that IP are now invalid.
-	if !e.ctCleaned {
+	if !e.ctCleaned && !owner.DryModeEnabled() {
```
Comment (raybejjani) [inline pkg/endpoint/bpf.go]: This seems more correct to me, now, but is it directly related to the other changes? I have no idea what dry mode is (seems like it wouldn't be any fun at a party) but if it's unrelated it might be better to have this in a separate commit, in case we broke something.

**Review by raybejjani — APPROVED** [review]
I think I understood this but, to be honest, I'm not sure what I don't know ;/ My review might be suboptimal

File: pkg/endpoint/bpf.go (hunk 3 above)
Comment (ianvernon) [reply pkg/endpoint/bpf.go]: `DryMode` is used for running unit-tests. I did this to clean up annoying errors that were clogging console while running tests locally. I'll factor this out into a separate PR since I want to get rid of `DryModeEnabled` access via `Owner` and grab it from `option.Config` anyway.

**Comment by ianvernon** [comment]
test-me-please

File: pkg/endpoint/policy.go (hunk 2 above)
Comment (ianvernon) [reply pkg/endpoint/policy.go]: Sigh, I missed this comment - will update the comment once build finishes, and will run gofmt locally and paste output to confirm I don't break master. Given that it's only a comment update, I won't run tests again provided I didn't break anything in my latest upload.

**Comment by ianvernon** [comment]
Prior build passed. Fixed comments and rebased against master; if passes, will merge.

**Comment by ianvernon** [comment]
test-me-please

**Comment by ianvernon** [comment]
Build failed due to infrastructural issues:
```
21:30:32 make[1]: go-bindata: Command not found
21:30:32 Makefile:54: recipe for target 'go-bindata' failed
```

Trying again.

**Comment by ianvernon** [comment]
test-me-please

File: pkg/endpoint/policy.go [hunk 4]
```diff
@@ -323,7 +319,7 @@ func (e *Endpoint) computeDesiredL3PolicyMapEntries(owner Owner, identityCache *
 
 	// Only L3 (label-based) policy apply.
 	// Complexity increases linearly by the number of identities in the map.
-	for identity, labels := range *identityCache {
+	for identity, labels := range *e.prevIdentityCache {
```
Comment (tgraf) [inline pkg/endpoint/policy.go:322]: I was a bit puzzled here at first. It's not obvious why prevIdentityCache is pointing to the right cache at first. Reading the rest of the code, it is but this may break as it's not obvious that the `e.prevIdentitycache` assignment cannot be moved around. I recommend to add a comment to the assignment to make sure we don't accidentally break this.
